File: bindings/dotnet/dotnet/StateSet/Crypto.cs
using System.Runtime.InteropServices;

namespace StateSet.Embedded;

/// <summary>
/// Cross-binding cryptographic primitives.
/// </summary>
/// <remarks>
/// Thin wrappers over the C-FFI exports in <c>bindings/dotnet/src/lib.rs</c>
/// that delegate to the <c>stateset-crypto</c> Rust crate. The same set of
/// primitives is exported from every StateSet binding and verified against
/// the language-neutral test corpus at
/// <c>bindings/test-vectors/v1.json</c>.
///
/// All methods are thread-safe and have no class state. On invalid input
/// or runtime failure they throw <see cref="InvalidOperationException"/>.
/// </remarks>
public static class Crypto
{
    /// <summary>
    /// Return the RFC 8785 JCS canonical-form bytes for a JSON string.
    /// </summary>
    /// <param name="json">JSON value as a UTF-8 string.</param>
    /// <returns>The canonical UTF-8 bytes.</returns>
    /// <exception cref="InvalidOperationException">
    /// Thrown if <paramref name="json"/> is invalid or canonicalization fails.
    /// </exception>
    public static byte[] JcsCanonicalize(string json)
    {
        var rc = NativeMethods.stateset_crypto_jcs_canonicalize(
            json,
            out IntPtr outPtr,
            out nuint outLen);
        if (rc != 0)
        {
            throw new InvalidOperationException(
                $"jcs_canonicalize failed (rc={rc}): invalid JSON or canonicalization error");
        }
        try
        {
            var managed = new byte[(int)outLen];
            Marshal.Copy(outPtr, managed, 0, (int)outLen);
            return managed;
        }
        finally
        {
            NativeMethods.stateset_crypto_free_buffer(outPtr, outLen);
        }
    }

    /// <summary>
    /// Compute the VES v1.0 payload-plain hash of a JSON payload.
    /// </summary>
    /// <param name="json">JSON payload as a UTF-8 string.</param>
    /// <param name="salt">16 bytes of salt, or <c>null</c> for unsalted.</param>
    /// <returns>The 32-byte digest.</returns>
    /// <exception cref="InvalidOperationException">
    /// Thrown if <paramref name="json"/> is invalid or <paramref name="salt"/> is
    /// not exactly 16 bytes.
    /// </exception>
    public static byte[] PayloadPlainHash(string json, byte[]? salt)
    {
        var output = new byte[32];
        IntPtr saltPtr = IntPtr.Zero;
        nuint saltLen = 0;
        GCHandle saltHandle = default;
        GCHandle outHandle = GCHandle.Alloc(output, GCHandleType.Pinned);
        try
        {
            if (salt != null)
            {
                if (salt.Length != 16)
                {
                    throw new InvalidOperationException(
                        $"salt must be exactly 16 bytes, got {salt.Length}");
                }
                saltHandle = GCHandle.Alloc(salt, GCHandleType.Pinned);
                saltPtr = saltHandle.AddrOfPinnedObject();
                saltLen = (nuint)salt.Length;
            }
            var rc = NativeMethods.stateset_crypto_payload_plain_hash(
                json,
                saltPtr,
                saltLen,
                outHandle.AddrOfPinnedObject());
            if (rc != 0)
            {
                throw new InvalidOperationException(
                    $"payload_plain_hash failed (rc={rc}): invalid JSON or hash error");
            }
            return output;
        }
        finally
        {
            if (saltHandle.IsAllocated) saltHandle.Free();
            outHandle.Free();
        }
    }

    /// <summary>
    /// Compute the merkle root of a list of 32-byte leaves.
    /// </summary>
    /// <param name="leaves">The leaf digests (each 32 bytes).</param>
    /// <returns>The 32-byte merkle root.</returns>
    /// <exception cref="InvalidOperationException">
    /// Thrown if any leaf is not exactly 32 bytes.
    /// </exception>
    public static byte[] MerkleRoot(byte[][] leaves)
    {
        var output = new byte[32];
        var outHandle = GCHandle.Alloc(output, GCHandleType.Pinned);
        try
        {
            if (leaves.Length == 0)
            {
                var rc0 = NativeMethods.stateset_crypto_merkle_root(
                    IntPtr.Zero,
                    0,
                    outHandle.AddrOfPinnedObject());
                if (rc0 != 0)
                {
                    throw new InvalidOperationException(
                        $"merkle_root failed for empty leaves (rc={rc0})");
                }
                return output;
            }
            var flat = new byte[leaves.Length * 32];
            for (int i = 0; i < leaves.Length; i++)
            {
                if (leaves[i].Length != 32)
                {
                    throw new InvalidOperationException(
                        $"merkle_root: leaf {i} must be 32 bytes, got {leaves[i].Length}");
                }
                Buffer.BlockCopy(leaves[i], 0, flat, i * 32, 32);
            }
            var flatHandle = GCHandle.Alloc(flat, GCHandleType.Pinned);
            try
            {
                var rc = NativeMethods.stateset_crypto_merkle_root(
                    flatHandle.AddrOfPinnedObject(),
                    (nuint)leaves.Length,
                    outHandle.AddrOfPinnedObject());
                if (rc != 0)
                {
                    throw new InvalidOperationException(
                        $"merkle_root failed (rc={rc})");
                }
                return output;
            }
            finally
            {
                flatHandle.Free();
            }
        }
        finally
        {
            outHandle.Free();
        }
    }
}

File: bindings/swift/Sources/StateSet/Crypto.swift
import Foundation
import StateSetC

/// Cross-binding cryptographic primitives.
///
/// Thin Swift wrappers over the C-FFI exports in `bindings/swift/src/lib.rs`
/// that delegate to the `stateset-crypto` Rust crate. The same primitives
/// are exported from every StateSet binding and verified against the
/// language-neutral test corpus at `bindings/test-vectors/v1.json`.
///
/// All methods are static and thread-safe. On invalid input or runtime
/// failure they throw `Crypto.Error`.
public enum Crypto {

    public enum Error: Swift.Error, CustomStringConvertible {
        case invalidInput(String)
        case operationFailed(String)

        public var description: String {
            switch self {
            case .invalidInput(let msg): return "invalid input: \(msg)"
            case .operationFailed(let msg): return "operation failed: \(msg)"
            }
        }
    }

    /// RFC 8785 JCS canonical-form bytes for a JSON string.
    public static func jcsCanonicalize(_ json: String) throws -> Data {
        var outPtr: UnsafeMutablePointer<UInt8>? = nil
        var outLen: size_t = 0
        let rc = json.withCString { cstr in
            stateset_crypto_jcs_canonicalize(cstr, &outPtr, &outLen)
        }
        if rc != 0 {
            throw Error.operationFailed("jcs_canonicalize rc=\(rc)")
        }
        guard let ptr = outPtr else {
            throw Error.operationFailed("jcs_canonicalize returned null pointer")
        }
        defer { stateset_crypto_free_buffer(ptr, outLen) }
        return Data(bytes: ptr, count: outLen)
    }

    /// VES v1.0 payload-plain hash. Returns 32 bytes.
    /// `salt`, when non-nil, must be exactly 16 bytes.
    public static func payloadPlainHash(_ json: String, salt: Data? = nil) throws -> Data {
        if let s = salt, s.count != 16 {
            throw Error.invalidInput("salt must be exactly 16 bytes, got \(s.count)")
        }
        var output = [UInt8](repeating: 0, count: 32)
        let rc: Int32 = json.withCString { cstr in
            output.withUnsafeMutableBufferPointer { outBuf in
                if let s = salt {
                    return s.withUnsafeBytes { saltRaw -> Int32 in
                        let saltPtr = saltRaw.bindMemory(to: UInt8.self).baseAddress
                        return stateset_crypto_payload_plain_hash(
                            cstr,
                            saltPtr,
                            s.count,
                            outBuf.baseAddress)
                    }
                } else {
                    return stateset_crypto_payload_plain_hash(
                        cstr,
                        nil,
                        0,
                        outBuf.baseAddress)
                }
            }
        }
        if rc != 0 {
            throw Error.operationFailed("payload_plain_hash rc=\(rc)")
        }
        return Data(output)
    }

    /// Merkle root of a list of 32-byte leaves. Returns 32 bytes.
    /// An empty list yields the empty-tree sentinel from `stateset-crypto`.
    public static func merkleRoot(_ leaves: [Data]) throws -> Data {
        var output = [UInt8](repeating: 0, count: 32)
        if leaves.isEmpty {
            let rc = output.withUnsafeMutableBufferPointer { outBuf in
                stateset_crypto_merkle_root(nil, 0, outBuf.baseAddress)
            }
            if rc != 0 {
                throw Error.operationFailed("merkle_root rc=\(rc) for empty leaves")
            }
            return Data(output)
        }
        var flat = [UInt8]()
        flat.reserveCapacity(leaves.count * 32)
        for (i, leaf) in leaves.enumerated() {
            if leaf.count != 32 {
                throw Error.invalidInput("leaf \(i) must be 32 bytes, got \(leaf.count)")
            }
            flat.append(contentsOf: leaf)
        }
        let rc: Int32 = flat.withUnsafeBufferPointer { flatBuf in
            output.withUnsafeMutableBufferPointer { outBuf in
                stateset_crypto_merkle_root(
                    flatBuf.baseAddress,
                    leaves.count,
                    outBuf.baseAddress)
            }
        }
        if rc != 0 {
            throw Error.operationFailed("merkle_root rc=\(rc)")
        }
        return Data(output)
    }
}
